Source/WebCore/platform/graphics/texmap/TextureMapperLayer3DRenderingContext.cpp

```diff
@@ -29,118 +29,300 @@
 {
-    if (layers.isEmpty())
+    if (textureMapperLayers.isEmpty())
         return;
 
-    Deque<TextureMapperLayerPolygon> layerList;
-    for (auto* layer : layers)
-        layerList.append({ { layer->effectiveLayerRect(), layer->toSurfaceTransform() }, layer, false });
-
-    auto root = makeUnique<TextureMapperLayerNode>(layerList.takeFirst());
-    buildTree(*root, layerList);
-
-    // Collect clip data
-    Vector<float> clipVertices;
-    traverseTree(*root, [&clipVertices](TextureMapperLayerNode& node) {
-        for (auto& polygon : node.polygons) {
-            auto toLayerTransform = polygon.layer->toSurfaceTransform().inverse();
-            if (polygon.isSplitted && toLayerTransform) {
-                polygon.clipVertexBufferOffset = clipVertices.size();
-
-                unsigned numVertices = polygon.geometry.numberOfVertices();
-                for (unsigned i = 0; i < numVertices; i++) {
-                    auto v = toLayerTransform->mapPoint(polygon.geometry.vertexAt(i));
-                    clipVertices.append(v.x());
-                    clipVertices.append(v.y());
-                }
+    Vector<Layer> layers;
+    for (size_t i = 0; i < textureMapperLayers.size(); ++i) {
+        auto layer = textureMapperLayers[i];
+        auto layerRect = layer->effectiveLayerRect();
+        auto transform = layer->toSurfaceTransform();
+        layers.append({ { layerRect, transform }, layer, false });
+    }
+
+    // Perform a broad-phase sweep-and-prune across all axes to identify potential intersections.
+    // By determining which layers might intersect, we can limit BSP cutting planes to those areas only,
+    // preventing unnecessary splitting of layers that are spatially distant.
+    // This optimization matters because the TextureMapper uses stencil operations to render split layers,
+    // and on some hardware, stencil usage is slow. Reducing unnecessary splits helps improve performance.
+    auto potentialIntersectionsXAxis = sweepAndPrune(layers, SweepAndPruneAxis::XAxis);
+    auto potentialIntersectionsYAxis = sweepAndPrune(layers, SweepAndPruneAxis::YAxis);
+    auto potentialIntersectionsZAxis = sweepAndPrune(layers, SweepAndPruneAxis::ZAxis);
+
+    // Filter intersections and take pairs that are present in all sweep and prune sets
+    SweepAndPrunePairs potentialIntersectionsXY;
+    std::set_intersection(
+        potentialIntersectionsXAxis.begin(), potentialIntersectionsXAxis.end(),
+        potentialIntersectionsYAxis.begin(), potentialIntersectionsYAxis.end(),
+        std::inserter(potentialIntersectionsXY, potentialIntersectionsXY.begin())
+    );
+    SweepAndPrunePairs potentialIntersections;
+    std::set_intersection(
+        potentialIntersectionsXY.begin(), potentialIntersectionsXY.end(),
+        potentialIntersectionsZAxis.begin(), potentialIntersectionsZAxis.end(),
+        std::inserter(potentialIntersections, potentialIntersections.begin())
+    );
```

Comment:
I think this is inefficient.
How about this?
* Adding `FloatBox3D::intersects(const FloatBox3D&)` method like `FloatRect::intersects`.
* Checking if each pair of layers intersect

Reply:
I was a bit too hasty in pushing my code. I accidentally pushed debug branch that had some extra. Code is now update. Doing sweep and prune on all axis is effectively same as doing bounding box intersection test. However, my idea is to do sweep and prune only on x-axis. This reduces the set needed for zy intersection tests. Updated the code to do only x axis sweep and prune

Reply:
Why only on x-axis? bounding box intersection test is pretty cheep.

Reply:
Yes, bounding box intersection tests are inexpensive. However, using them here would result in duplicate tests since the zy intersection already accounts for bounding box scenarios along the z and y dimensions. Additionally, full bounding box tests would involve testing each layer against every other layer, leading to redundant computations. An x-axis sweep-and-prune efficiently reduces the number of pairwise tests. That said, for typical scene sizes, the performance gain over full bounding box tests might be negligible.